hudi-common/src/main/java/org/apache/hudi/common/model/HoodieRecord.java

```diff
@@ -211,7 +213,14 @@ public HoodieOperation getOperation() {
     return operation;
   }
 
-  public abstract Comparable<?> getOrderingValue(Schema recordSchema, Properties props);
+  public Comparable<?> getOrderingValue(Schema recordSchema, Properties props) {
+    if (orderingValue == null) {
+      orderingValue = doGetOrderingValue(recordSchema, props);
+    }
+    return orderingValue;
+  }
+
+  protected abstract Comparable<?> doGetOrderingValue(Schema recordSchema, Properties props);
```

Comment:
could u add java doc to explain the difference between `doGetOrderingValue` and `getOrderingValue` to avoid confusing. maybe we can rename to something like `extractOrderingValueFromData`?